### PR TITLE
Vectorize _create_custom_4d_mask in CustomQwen2Decoder

### PR DESCRIPTION
**Pull request by ckvermaAI**
## Motivation

`_create_custom_4d_mask` inside `CustomQwen2ModelInner` built the `[B, 1, S, S]` attention mask with a Python-level double loop — one outer loop over batch items and one inner `for i, text_pos in enumerate(text_positions)` loop over text token positions. On GPU this is pathologically slow: each iteration synchronises the device to resolve the scalar Python index, and the number of kernel launches grows as `O(B × S_text)`.

Measured on A100 (CUDA events, 50 iterations):

| B | S | Before | After | Speedup |
|---|---|--------|-------|---------|
| 1 | 256 | 17.3 ms | 0.21 ms | **84×** |
| 4 | 512 | 137 ms | 0.21 ms | **655×** |
| 8 | 1024 | 549 ms | 0.27 ms | **2035×** |
| 16 | 2048 | 2196 ms | 1.93 ms | **1137×** |
| 4 | 4096 | 551 ms | 2.01 ms | **274×** |

## Modifications

Replace the Python loops with three fully-batched boolean tensor operations that encode the same three attention rules:

| Rule | Before | After |
|------|--------|-------|
| Image → all image (bidirectional) | scatter over `image_positions[:, None]` per batch item | `img_outer = is_image.unsqueeze(2) & is_image.unsqueeze(1)` — `[B,S,S]` outer product |
| Text → all image | scatter per text token per batch item | `text_to_img = is_text.unsqueeze(2) & is_image.unsqueeze(1)` |
| Text → causal text (j ≤ i) | inner `enumerate` loop | `text_causal = is_text[:,:,None] & is_text[:,None,:] & causal[None]` where `causal[q,k] = (k≤q)` is a single `torch.arange` comparison |

The three masks are OR-ed into a single `allow` tensor and applied in one `mask[allow] = 0.0` scatter, then `unsqueeze(1)` adds the head dimension. The output shape `[B, 1, S, S]` and all mask values (`0.0` for allowed, `min_dtype` for blocked) are identical to the original.

### Tests
New file `test/manual/test_create_custom_4d_mask.py`:

- **Accuracy (17 cases)** — bit-exact comparison of new vs reference across: all-image, all-text, image-then-text, single image/text token, random interleaving, heterogeneous batch, `float32`/`float16`/`bfloat16`. Semantic checks verify causal ordering, bidirectional image attention, and text→image attention individually.
- **Performance (5 cases)** — asserts new impl is not slower than reference across the benchmark sweep above. On CUDA uses `torch.cuda.Event` timing for precise GPU measurement; falls back to `perf_counter` on CPU.
- **PyTorch profiler** — `--profile` flag (or `PROFILE_TRACES=1`) captures Chrome-trace JSON for both implementations with `ProfilerActivity.CPU` + `ProfilerActivity.CUDA` (when available). Each implementation's warmup and measured iterations are wrapped in `record_function` spans so they are clearly distinguished in chrome://tracing.

```
# run accuracy + perf on CUDA:
python test/manual/test_create_custom_4d_mask.py -v

# capture profiler traces:
python test/manual/test_create_custom_4d_mask.py --profile --trace-dir ./pt_traces
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #27752737573](https://github.com/sgl-project/sglang/actions/runs/27752737573)
Latest PR Test (Extra): :x: [Run #27932210197](https://github.com/sgl-project/sglang/actions/runs/27932210197)

### CODE REVIEW

**Comment by ckvermaAI** [comment]
@mingfeima Please review this PR

**Comment by ckvermaAI** [comment]
@baonudesifeizhai Please review this PR

**Comment by mingfeima** [comment]
@jianan-gu please help review this one!

**Comment by ckvermaAI** [comment]
Addressed the review comments from co-pilot. Unit test reimplements the _create_custom_4d_mask since it's a method of nested local class.

**Review by mingfeima — COMMENTED** [review]
does CI has test cases to cover correctness of this change? the new test case is in manual not registered.

**Review by mingfeima — APPROVED** [review]

**Comment by yuan-luo** [comment]
This PR doesn't pass lint.
